OpenSimplex data optimization

nevarek opened the pull request:
<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->

# What Changed?
The images created in `Image::get_texture()` and `Image::get_seamless_image()` were changed from `FORMAT_RGBA8` to `FORMAT_L8`. As a result, the code responsible for allocating and generating values for the redundant memory has been removed.

# Explanation:
For every data point (pixel) in OpenSimplex images:
- Alpha is always 255
- Red == Green == Blue

The Alpha and redundant RGB data can be safely omitted without losing noise data used for the generated image.

---

Discussion source: https://www.reddit.com/r/godot/comments/klawh0/opensimplexnoiseget_image_overhead_intended/
Credit should go to /u/TheHumbleDM for finding/squashing this one!

Xrayez commented:
Yes, that's what I'm doing in some of my other noise-generator module in https://github.com/Xrayez/godot-anl/blob/642b470e8c1e1943e5c8016be2fc28213d5702b8/noise.cpp#L880-L901 (for heightmap/bumpmap only).

For using the noise as a normal map, this still requires RGBA8 format, but I don't know whether normal map generation in `NoiseTexture` depends on the source format in `OpenSimplexNoise`. Does it still work if you set `NoiseTexture.as_normalmap` property on? Should likely work regardless but worth to ask anyways...

Calinou reviewed (approved):
I tested this locally, it works fine. Normal map conversion still works:

![image](https://user-images.githubusercontent.com/180032/103467652-75dc6580-4d51-11eb-8d7d-5c78134f7514.png)

akien-mga commented:
Could you squash the commits into one?

nathansmith339 commented:
I'm pretty sure I squashed it correctly, let me know if it's not.

akien-mga commented:
Could you also mention this in the documentation for `get_image()`?
E.g.:
```diff
diff --git a/modules/opensimplex/doc_classes/OpenSimplexNoise.xml b/modules/opensimplex/doc_classes/OpenSimplexNoise.xml
index 9fe4c9c249..3c1d31e9a3 100644
--- a/modules/opensimplex/doc_classes/OpenSimplexNoise.xml
+++ b/modules/opensimplex/doc_classes/OpenSimplexNoise.xml
@@ -32,7 +32,7 @@
                        <argument index="1" name="height" type="int">
                        </argument>
                        <description>
-                               Generate a noise image with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
+                               Generate a noise image in format [constant Image.FORMAT_L8] with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
                        </description>
                </method>
                <method name="get_noise_1d" qualifiers="const">
```

JFonS reviewed (approved):
Changes look good.

Note that in order to cherry-pick this into 3.2, this line https://github.com/godotengine/godot/blob/cc5c56c4dff453f01154ac48d1dd958b46588dbb/modules/opensimplex/noise_texture.cpp#L104

Will need to be changed so `Image::FORMAT_RGBA8` becomes `p_image->get_format()`.

nathansmith339 commented:
I updated the documentation for both `get_image` and `get_seamless_image` to reflect the format changes

akien-mga commented on modules/opensimplex/doc_classes/OpenSimplexNoise.xml:
```diff
@@ -32,7 +32,7 @@
 			</argument>
 			<description>
-				Generate a noise image with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
+				Generate a noise image in format [constant Image.FORMAT_L8] with the requested [code]width[/code] and [code]height[/code], based on the current noise parameters.
```
I'd use `in [constant Image.FORMAT_L8] format` as done for the other method, to be consistent in the language.

nathansmith339 replied:
Thanks! I was debating this, too. I'll err on the side of consistency from now on.

nevarek committed:
Optimize data format for OpenSimplex images

The previous RGBA format included unused RGB data. Using the LA8 format
removes the need to store the extra data.

The Docs have been updated to reflect the format changes.

nathansmith339 commented:
I changed the wording to be consistent.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:

akien-mga commented:
Cherry-picked for 3.2.4.
